circuit_breaker.go

```diff
@@ -0,0 +1,159 @@
+package resty
+
+import (
+	"errors"
+	"net/http"
+	"sync/atomic"
+	"time"
+)
+
+// CircuitBreaker can be in one of three states: Closed, Open, or Half-Open.
+// When the circuit breaker is Closed, requests are allowed to pass through.
+// If a failure count threshold is reached within a specified time-frame, the circuit breaker transitions to the Open state.
+// When the circuit breaker is Open, requests are blocked.
+// After a specified timeout, the circuit breaker transitions to the Half-Open state.
+// When the circuit breaker is Half-Open, a single request is allowed to pass through.
+// If that request fails, the circuit breaker returns to the Open state.
+// If SuccessThreshold requests succeed, the circuit breaker transitions back to the Closed state.
+type CircuitBreaker struct {
+	policies                        []CircuitBreakerPolicy
+	timeout                         time.Duration
+	failThreshold, successThreshold int
+
+	state                   circuitBreakerState
+	failCount, successCount int
+	lastFail                time.Time
+}
+
+// NewCircuitBreaker creates a new CircuitBreaker with default settings.
+// The default settings are:
+// - Timeout: 10 seconds
+// - FailThreshold: 3
+// - SuccessThreshold: 1
+// - Policies: Count5xxPolicy
+func NewCircuitBreaker() *CircuitBreaker {
+	return &CircuitBreaker{
+		policies:         []CircuitBreakerPolicy{Count5xxPolicy},
+		timeout:          10 * time.Second,
+		failThreshold:    3,
+		successThreshold: 1,
+	}
+}
+
+// SetPolicies sets the CircuitBreakerPolicy's that the CircuitBreaker will use to determine whether a response is a failure.
+func (cb *CircuitBreaker) SetPolicies(policies []CircuitBreakerPolicy) *CircuitBreaker {
+	cb.policies = policies
+	return cb
+}
+
+// SetTimeout sets the timeout duration for the CircuitBreaker
+func (cb *CircuitBreaker) SetTimeout(timeout time.Duration) *CircuitBreaker {
+	cb.timeout = timeout
+	return cb
+}
+
+// SetFailThreshold sets the number of failures that must occur within the timeout duration for the CircuitBreaker to
+// transition to the Open state.
+func (cb *CircuitBreaker) SetFailThreshold(threshold int) *CircuitBreaker {
+	cb.failThreshold = threshold
+	return cb
+}
+
+// SetSuccessThreshold sets the number of successes that must occur to transition the CircuitBreaker from the Half-Open state
+// to the Closed state.
+func (cb *CircuitBreaker) SetSuccessThreshold(threshold int) *CircuitBreaker {
+	cb.successThreshold = threshold
+	return cb
+}
+
+// CircuitBreakerPolicy is a function that determines whether a response should trip the circuit breaker
+type CircuitBreakerPolicy func(resp *http.Response) bool
+
+// Count5xxPolicy is a CircuitBreakerPolicy that trips the circuit breaker if the response status code is 500 or greater
+func Count5xxPolicy(resp *http.Response) bool {
+	return resp.StatusCode > 499
+}
+
+var ErrCircuitBreakerOpen = errors.New("rest: circuit breaker open")
```

Comment:
This, no?

```suggestion
var ErrCircuitBreakerOpen = errors.New("resty: circuit breaker open")
```

Or I don't get the  `rest:` maybe

Reply:
Done